### ISSUE TITLE
Dynamic mapping is not inherited from parent object

### ISSUE THREAD

**Issue by gilad61** [elastic/elasticsearch#22010]
**Elasticsearch version**: 2.4

**JVM version**: 1.8.0_102

**OS version**: Windows 10

**Description of the problem including expected versus actual behavior**:
According to [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/2.4/dynamic.html), dynamic field setting should be inherited from the parent object or from the top level type.
However, when a parent field has 'dynamic: false', the child fields do not honor this setting.

**Steps to reproduce**:
 1. Create a mapping with 'dynamic: false' object and some child field
```
PUT http://localhost:9200/testindex/_mapping/sometype
{
  "properties": {
    "data": {
      "dynamic": false,
      "properties": {
        "child": {
          "properties": { "name": { "type": "string" } }
        }
      }
    }
  }
}
```
 2. Add a document with some non-mapped field under the 'child' field
```
POST http://localhost:9200/testindex/sometype/1
{
  "data": { 
    "child": { 
      "newField": "I shouldn't be here" 
    } 
  }
}
```
 3. Check the index mapping and see that 'newField' was added dynamically
```
GET http://localhost:9200/testindex/_mapping
```
Result:
```
"sometype" : {
  "properties" : {
    "data" : {
      "dynamic" : "false",
      "properties" : {
        "child" : {
          "properties" : {
            "name" : { "type" : "string" },
            "newField" : { "type" : "string" }
          }
        }
      }
    }
  }
}
```

**Comment by clintongormley** [comment]
Hi @gilad61 

This is a known bug in 2.x, and is fixed in 5.0.  We aren't able to backport the fix I'm afraid.

**Comment by ntvo** [comment]
i think you should set "dynamic":"strict" instead.
dynamic set to false only make the field not index, sourceid still add it.
Nathan Vo
